cmd/kubeapps-apis/server/server.go

```diff
@@ -59,28 +53,40 @@ func Serve(serveOpts ServeOptions) error {
 	if err != nil {
 		return fmt.Errorf("Failed to create gateway: %v", err)
 	}
-	gwArgs := gwHandlerArgs{
-		ctx:         ctx,
-		mux:         gw,
-		addr:        listenAddr,
-		dialOptions: []grpc.DialOption{grpc.WithInsecure()},
+	gwArgs := core.GatewayHandlerArgs{
+		Ctx:         ctx,
+		Mux:         gw,
+		Addr:        listenAddr,
+		DialOptions: []grpc.DialOption{grpc.WithInsecure()},
 	}
 
-	// Create the core.plugins server which handles registration of plugins,
-	// and register it for both grpc and http.
-	pluginsServer, err := NewPluginsServer(serveOpts, grpcSrv, gwArgs)
+	// Create the core.plugins.v1alpha1 server which handles registration of
+	// plugins, and register it for both grpc and http.
+	pluginsServer, err := pluginsv1.NewPluginsServer(serveOpts, grpcSrv, gwArgs)
 	if err != nil {
 		return fmt.Errorf("failed to initialize plugins server: %v", err)
 	}
 	plugins.RegisterPluginsServiceServer(grpcSrv, pluginsServer)
-	err = plugins.RegisterPluginsServiceHandlerFromEndpoint(gwArgs.ctx, gwArgs.mux, gwArgs.addr, gwArgs.dialOptions)
+	// TODO(minelson): should no longer need to split up the gw args now.
+	err = plugins.RegisterPluginsServiceHandlerFromEndpoint(gwArgs.Ctx, gwArgs.Mux, gwArgs.Addr, gwArgs.DialOptions)
 	if err != nil {
 		return fmt.Errorf("failed to register core.plugins handler for gateway: %v", err)
 	}
 
+	// Ask the plugins server for plugins with GRPC servers that fulfil the core
+	// packaging v1alpha1 API, then pass to the constructor below.
+	// The argument for the reflect.TypeOf is based on what grpc-go
+	// does itself at:
+	// https://github.com/grpc/grpc-go/blob/v1.38.0/server.go#L621
+	packagingPlugins := pluginsServer.GetPluginsSatisfyingInterface(reflect.TypeOf((*packages.PackagesServiceServer)(nil)).Elem())
```

Comment:
Nice, thanks for the reference. However, I'm wondering: if a given pkg plugin has multiple versions that satisfy the core API (ex: plugin v1: core ops, plugin v2: core ops + customs ops)... would these two versions be added here?
If so, we would be aggregating repeated results, wouldn't we?

Reply:
Yes, if the api server is serving the core.packages.v1alpha1 api only, then you would want to enable only a single (for example) helm plugin that satisfies that API (so in your example, I think there would be no need for the `v1` one if the `v2` one was enabled as both implement the core ops). Or if that becomes an issue, we can enforce some restrictions on startup (such as aggregating only one / the latest / whatever, or just refusing to start if more than one per packaging system is found to implement it).

The purpose is that we later may serve both a `core.packages.v1` API and a `core.packages.v2` API with a separate helm plugin for each... which is easy as long as the `v2` API doesn't satisfy the `v1` API (which it should not, and we can enforce that, I think).